cmd/faFindFast/efficient.go

```diff
@@ -4,133 +4,210 @@ import (
+// outputs:
+//  1. alnIdx: the next alignment position where a firstQuery base exists, which is usually alignmentIndex+1, but can be greater due to gaps in the firstQuery sequence
+//     this value will be equal to len(firstQuery) if we were unable to find another firstQuery base, and this would mean we have reached the end of the firstQuery sequence
+//  2. gapOpenCloseFirstQuery: gaps opened in the firstQuery sequence (0 or 1)
+//  3. gapOpenedSecondQuery: gaps opened in the secondQuery sequence (0 or 1)
+//  4. gapClosedSecondQuery: gaps closed in the secondQuery sequence (0 or 1)
+//  5. numFirstQueryNs: the number of Ns in the firstQuery after this increment (0 or 1)
+//  6. numSecondQueryNsGap: the number of Ns in the secondQuery sequence after this increment from gaps in the firstQuery
+//  7. numSecondQueryNsMatch: the number of Ns in the secondQuery sequence after this increment from matches with the firstQuery
+//  8. numSubst: the number of substitutions/mismatches (0 or 1) between firstQuery and secondQuery
+func incrementWindowEdge(firstQuery []dna.Base, secondQuery []dna.Base, alnIdxOrig int) (alnIdx, gapOpenCloseFirstQuery, gapOpenedSecondQuery, gapClosedSecondQuery, numFirstQueryNs, numSecondQueryNsGap, numSecondQueryNsMatch, numSubst int) {
 	alnIdx = alnIdxOrig
-
-	// increment alnIdx to the next ref position (next non-gap pos in seqOne)
-	for alnIdx++; alnIdx < len(seqOne) && seqOne[alnIdx] == dna.Gap; alnIdx++ {
-		if seqTwo[alnIdx] == dna.N {
-			numQueryNsGap++
+	// at initialization, lastAlnIdxOfWindow == -1, which is assigned to alnIdxOrig when incrementWindowEdge is called, so alnIdx = alnIdxOrig == -1. For loop will start at alnIdx++ which is 0, the 1st position in the alignment
+
+	// increment alnIdx to the next non-gap position in the firstQuery sequence
+	// for loop start condition: start at alnIdx++, which is the position after the current alignment position
+	// for loop end condition: ends when we reach the end of the firstQuery sequence, or when we reach a non-gap position in the firstQuery sequence, whichever happens first
+	// the for loop's start and end conditions also mean that the for loop is never executed, unless there is firstQuery gap before we reach the end of firstQuery
+	// for loop incrementation: alnIdx++
+	for alnIdx++; alnIdx < len(firstQuery) && firstQuery[alnIdx] == dna.Gap; alnIdx++ {
+		// keep track of changes between the original and next non-gap positions in the firstQuery sequence
+		// 6. secondQuery
+		if secondQuery[alnIdx] == dna.N {
+			numSecondQueryNsGap++
 		}
-		if seqTwo[alnIdx] != dna.Gap {
-			gapOpenCloseRef = 1
+		// 2. gapOpenClosedFirstQuery
+		if secondQuery[alnIdx] != dna.Gap {
+			gapOpenCloseFirstQuery = 1 // Rationale: if the for loop is executed, this means there is a gap in the firstQuery, and if secondQuery does not have a gap, then it is indeed a firstQuery gap compared to secondQuery
```

Comment:
This file has the biggest changes. Most of the changes in this file are renaming anything to do with "reference" to "firstQuery", which does not impact code efficiency.